### DISCUSSION TITLE
QUTE Concurrent loading of CDI beans.

### DISCUSSION THREAD

**Discussion by TonyWeston** [quarkusio/quarkus#30253] [Q&A] [answered]
Imagine I am working with a complex page, with many tags / components.  Imagine these tags need some time to resolve their data.  What is the best way of allowing those tags to build their data concurrently?   Let me show you an example of what I have so far, with just 2 tags.    

This is based on the QUTE example.  I have a page with 2 tags.  comp1 and comp2.

```
<body>
<h1>Hello <b>{name ?: "Qute"}</b></h1>

<p>Create your web page using Quarkus RESTEasy & Qute</p>

{#comp1/} <br/>
{#comp2/} <br/>

</body>
```

Comp1 and Comp2 both make use of a injected CDI bean, C1 and C2.  (Comp1 below. Comp2 is identical).     

```
Hello this is comp1

{inject:c1.val}
```

C1 and C2 beans look like the following.    Note there is a Thread.sleep(1000), to simulate the bean getting data it requires, from say a remote server.    

```
@RequestScoped
@Named("c1")
public class C1 {
	
CompletableFuture<String> val = new CompletableFuture<String>();
    @Inject
	ManagedExecutor exec;
    public static long startTime;
	
	@PostConstruct
	public void init() throws InterruptedException {
		System.out.println("init c1");
		startTime=System.currentTimeMillis();

		exec.submit(()->{
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			val.complete("This is C1");
		});
	}


	public String getVal() {
		System.out.println("getVal C1");
		return val.join();
	}
}
```

There is some code in there for timing how long this takes.  When I show the page I get:     

<img width="459" alt="Screenshot 2023-01-09 at 09 26 52" src="https://user-images.githubusercontent.com/46684275/211276597-1bf1bb3e-5465-4906-97ba-c16cd810c152.png">

On the console I get:     

<img width="355" alt="Screenshot 2023-01-09 at 09 27 32" src="https://user-images.githubusercontent.com/46684275/211276736-cdc6c180-93da-4f35-96d8-8d2f15811cb5.png">

Note that its taking about 2 seconds to build.    

So what is happening is both C1 and C2 beans are being instantiated only at the very moment QUTE requests val.  Because of this, it is impossible for both beans to be resolve the data they require concurrently. - Its like QUTE says, "Create C1, and give me Val1", and blocks while this happens, then it says "Create C2 and give me Val2", and blocks while this happens.      

What I would really like is for QUTE to resolve and instantiate every bean first,  and then pull the data from those beans only after all beans have been instanciated.   I can simulate this by in my controller, inject these beans, and call toString on them, to force them to be realised.    

```
@Path("/some-page")
public class SomePage {

    private final Template page;
    
    private final C1 c1;
    private final C2 c2;


    public SomePage(Template page, C1 c1, C2 c2) {
        this.page = requireNonNull(page, "page is required");
        this.c1=c1;
        this.c2=c2;
    }

    @GET
    @Produces(MediaType.TEXT_HTML)
    public TemplateInstance get(@QueryParam("name") String name) {
    	c1.toString();
    	c2.toString();
    	
        return page.data("name", name);
    }

}
```

Once this is done, the console says this:     

<img width="282" alt="Screenshot 2023-01-09 at 09 33 38" src="https://user-images.githubusercontent.com/46684275/211277875-98a64da0-e3e8-4d1c-a217-23d68f556f48.png">

Can you see how the time to show the page, is now 1 second, rather than 2 seconds before? This is because the controller is forcing these components to be instantiated up front, without requesting the data. This allows them to be built concurrently. Though this is my preferred behaviour,  I really do not like having to put references to the beans in my controller, just to force them to build.     

Is there an alternative pattern I should be using, to give me the same asynchronous behaviour, which will allow me to build a screen where the data for many tags, is built concurrently?

[a comment by an automated account is not shown]

**Comment by mkouba** [comment, marked as answer]
You should just return the `CompletableFuture` from the `getVal()` method, i.e. don't block when the `{inject:c1.val}` is evaluated.

**Reply by TonyWeston** [reply]
Tried it, Works great!  I had no idea that Qute could handle CompletableFuture.  Thank you.

**Reply by mkouba** [reply]
It does :smile: 

https://quarkus.io/guides/qute-reference#evaluation-of-completionstage-and-uni-objects
